tests/unit/migrations.py

```diff
@@ -0,0 +1,185 @@
+
+    `filename` is path to the SQL dump to load, relative to `pg_dump_dir`.
+
+    The SQL in that file will be executed, and then migration scripts will
+    be run to bring it up to date.
+    """
+    with open(path.join(pg_dump_dir, filename)) as f:
+        sql = f.read()
+    with app.app_context():
+        db.session.execute(sql)
+        db.session.commit()
+    upgrade(revision='heads')
+
+
+def fresh_create():
+    """Create a fresh database, and populate it with an initial set of objects.
+
+    The objects are created via `initial_db`. These objects should be such that
+    a migrated database dump will have the same contents.
+    """
+    create_db()
+    initial_db()
+
+
+def get_db_state():
+    """Inspect the database, and return a representation of its contents
+
+    The return value is a dictionary mapping table names to dictionaries
+    containing two keys each:
+
+        - 'schema', which describes the columns for that table.
+        - 'rows', which is a list of all rows in the table.
+    """
+    result = {}
+    inspector = db.inspect(db.engine)
+    metadata = db.MetaData()
+
+    for name in inspector.get_table_names():
+        schema = inspector.get_columns(name)
+        tbl = db.Table(name, metadata)
+        inspector.reflecttable(tbl, None)
+        rows = db.session.query(tbl).all()
+        result[name] = {
+            'schema': schema,
+            'rows': sorted(rows),
+        }
+
+    return result
+
+
+@pytest.mark.parametrize('filename,extra_config', [
+    ['flask.sql', {
+        'extensions': {
+            'haas.ext.switches.mock': '',
+            'haas.ext.switches.nexus': '',
+            'haas.ext.switches.dell': '',
+            'haas.ext.obm.ipmi': '',
+            'haas.ext.obm.mock': '',
+            'haas.ext.auth.database': '',
+            'haas.ext.network_allocators.vlan_pool': '',
+
+            # These are on by default; disable them.
+            'haas.ext.auth.null': None,
+            'haas.ext.network_allocators.null': None,
+        },
+        'haas.ext.network_allocators.vlan_pool': {
+            'vlans': '100-200, 300-500',
+        },
+    }],
+])
+def test_db_eq(filename, extra_config):
+    """Verify that each function in fns creates the same database."""
+
+    config_merge(extra_config)
+    load_extensions()
+    server.register_drivers()
+    server.validate_state()
+
+    def run_fn(f):
+        """Run the function f and return a representation of its db."""
+
+        # We can't just do db.drop_all(), since db's metadata won't
+        # necessarily reflect the contents of the database. If there are
+        # tables it doesn't know about, it may raise an exception.
+        with app.app_context():
+            drop_tables()
+            f()
+
+        return get_db_state()
+
+    upgraded = run_fn(lambda: load_dump(filename))
+    fresh = run_fn(fresh_create)
+    drop_tables()
+
+    def censor_nondeterminism(string):
```

Comment:
Quoting henn (2016-04-05 01:20:23)

>    What about adding an argument to Headnode.__init__() that takes a
>    uuid with a default of generating a new uuid?

The current solution is more extensible to other potential sources of
non-determinism, and doesn't require special support in the non-test
code. Especially the latter part of that I like, since it avoids
complicating other interfaces. However, if you have a strong opinion
here I'll defer; let me know.

>    Also, if you do like that approach, please change the default to use
>    uuid.uuid4() instead of uuid.uuid1() so that we get some
>    randomness.

I don't know that the randomness actually does much for us here --- it
would be perilous to assume uuid4 is *cryptographically* random (I
imagine it's not), and otherwise any mechanism for avoiding accidental
collisions is just as good, no? Not that I'm *opposed* to the random
one, but I'm dubious of the motivation.

Right now we're only using it for having unique libvirt names; we could
just use the primary key for that. That would require renaming all the
VMs though, which is probably not worth the trouble.